### ISSUE TITLE
GPU OOM

### ISSUE THREAD

**Issue by lelayf** [ibab/tensorflow-wavenet#53]
Running train.py on Nvidia GRID K2, 1536 cores, 3.5 GB memory.

`I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 0 with properties: 
name: GRID K2
major: 3 minor: 0 memoryClockRate (GHz) 0.745
pciBusID 0000:00:04.0
Total memory: 3.50GiB
Free memory: 3.45GiB
I tensorflow/core/common_runtime/gpu/gpu_init.cc:126] DMA: 0 
I tensorflow/core/common_runtime/gpu/gpu_init.cc:136] 0:   Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:838] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GRID K2, pci bus id: 0000:00:04.0)
Trying to restore saved checkpoints from ./logdir/train/2016-09-20T19-23-04 ... No checkpoint found.
step 0 - loss = 7.461, (13.180 sec/step)
Storing checkpoint to ./logdir/train/2016-09-20T19-23-04 ... Done.`

Then the BFC allocator runs out of memory:

`W tensorflow/core/common_runtime/bfc_allocator.cc:270] **************************************************************************************__**********xx
W tensorflow/core/common_runtime/bfc_allocator.cc:271] Ran out of memory trying to allocate 194.79MiB.  See logs for memory state.
W tensorflow/core/framework/op_kernel.cc:940] Resource exhausted: OOM when allocating tensor with shape[199462,256]
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:244] PoolAllocator: After 3463 get requests, put_count=3233 evicted_count=1000 eviction_rate=0.30931 and unsatisfied allo
cation rate=0.38406`

Do you know if there is a setting I can specify to work around GPU memory limitations ?

**Comment by larsr** [comment]
Are you running something else already on the GPU?

**Comment by lelayf** [comment]
nope, fresh install, fresh reboot, running headless and only process using the GPU.

**Comment by ibab** [comment]
Looks like it stopped on 2 particularly large input samples.
Did this happen immediately, or after a few iterations?
If it's just because of large samples, pull request #54 should fix this by always cutting them into fixed size pieces.

You can also try setting the `BATCH_SIZE` to 1, or reducing the number of dilated convolutional layers in `wavenet_params.json`.

3.5 GB isn't a lot, but should be enough to train the network with some tweaks.

Note that we haven't yet managed to reproduce the results from the wavenet paper, so it's not worth training the network unless you can invest the time to look for good hyperparameters.

**Comment by tuba** [comment]
I get OOM on 8GB video ram always after 160 step

**Comment by lelayf** [comment]
BATCH_SIZE=1 allows the process to run for a few dozen steps before OOMing. I just tried to mount a second identical GPU to increase total mem to 7GB but it behaves similarly. The second GPU is well recognized by TF but I think we would need explicit placement to make it useful. I will now look into #54

**Comment by ibab** [comment]
@lelayf: Yeah, we're not making use of extra GPUs at the moment.
#54 should definitely help. You can adjust the `SAMPLE_SIZE` downwards if you still run into problems.

**Comment by lelayf** [comment]
it seems the training is now stalling silently after a few hundred steps. I do not get OOMs. I tried using 96000 sample_size, then 64000 and finally ran it with no sample_size on the command line. I also tried different GPUs, a GRID K2 and a Tesla K80. In all cases the same silent stalling arises.

**Comment by ibab** [comment]
Could this be the same problem as in #65?
The audio pipeline stopped processing data after traversing the input files once.
Updating to one of the newer commits should fix that.

**Comment by ibab** [comment]
Closing this, as all mentioned issues should be fixed at this point.
Feel free to comment if you still experience problems with this.

**Comment by mikelibg** [comment]
Getting same OOM when running on 61GiB aws instance even with sample size 10,000.
commit: 3c973c038c8c2c20fef0039f111cb04139ff594b
python: 3.6
with --silence_threshold=0 on one utterance p225 from the original VCTK corpus

```
>>> psutil.virtual_memory()
svmem(total=64389132288, available=60933558272, percent=5.4, used=2929561600, free=59257782272, active=3300839424, inactive=1334075392, buffers=73183232, cached=2128605184, shared=21499904, slab=143368192)
```

Exception:
2018-05-13 06:20:26.790439: W tensorflow/core/framework/op_kernel.cc:1318] OP_REQUIRES failed at slice_op.cc:154 : Resource exhausted: OOM when allocating tensor with shape[1,44050,256] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
Storing checkpoint to ./logdir/train/2018-05-13T06-19-22 ... Done.
Traceback (most recent call last):
  File "/home/ubuntu/training/tensorflow-wavenet/venv/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1322, in _do_call
    return fn(*args)
  File "/home/ubuntu/training/tensorflow-wavenet/venv/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1307, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/home/ubuntu/training/tensorflow-wavenet/venv/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1409, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.ResourceExhaustedError: OOM when allocating tensor with shape[1,38934,256] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
	 [[Node: wavenet_1/loss/Slice = Slice[Index=DT_INT32, T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:GPU:0"](wavenet_1/loss/Reshape, wavenet_1/loss/Slice/begin, wavenet_1/loss/Slice/size)]]
